File: Estruturais/Decorator.py
# Pizza base
class Pizza:
    def descricao(self):
        return "Pizza simples"

# Decorador base
class DecoradorPizza(Pizza):
    def __init__(self, pizza):
        self.pizza = pizza

    def descricao(self):
        return self.pizza.descricao()

# Decorador que adiciona queijo extra
class QueijoExtra(DecoradorPizza):
    def descricao(self):
        return self.pizza.descricao() + " com queijo extra"

# Decorador que adiciona cogumelos
class Cogumelos(DecoradorPizza):
    def descricao(self):
        return self.pizza.descricao() + " com cogumelos"

# Exemplo de uso
pizza = Pizza()
pizza_com_queijo = QueijoExtra(pizza)
pizza_com_queijo_e_cogumelos = Cogumelos(pizza_com_queijo)
print(pizza_com_queijo_e_cogumelos.descricao())  # Pizza simples com queijo extra com cogumelos

File: Estruturais/Decorator.java
package Estruturais;

// Interface Pizza
interface Pizza {
    String descricao();
}

// Pizza simples
class PizzaSimples implements Pizza {
    public String descricao() {
        return "Pizza simples";
    }
}

// Decorador base
class DecoradorPizza implements Pizza {
    protected Pizza pizza;

    public DecoradorPizza(Pizza pizza) {
        this.pizza = pizza;
    }

    public String descricao() {
        return pizza.descricao();
    }
}

// Decorador queijo extra
class QueijoExtra extends DecoradorPizza {
    public QueijoExtra(Pizza pizza) {
        super(pizza);
    }

    @Override
    public String descricao() {
        return pizza.descricao() + " com queijo extra";
    }
}

// Decorador cogumelos
class Cogumelos extends DecoradorPizza {
    public Cogumelos(Pizza pizza) {
        super(pizza);
    }

    @Override
    public String descricao() {
        return pizza.descricao() + " com cogumelos";
    }
}

// Exemplo de uso
public class Decorator {
    public static void main(String[] args) {
        Pizza pizza = new PizzaSimples();
        Pizza pizzaComQueijo = new QueijoExtra(pizza);
        Pizza pizzaComQueijoECogumelos = new Cogumelos(pizzaComQueijo);

        System.out.println(pizzaComQueijoECogumelos.descricao());  // Pizza simples com queijo extra com cogumelos
    }
}


